clang/include/clang/Basic/AttributeCommonInfo.h

```diff
@@ -186,14 +186,14 @@ class AttributeCommonInfo {
   bool isGNUScope() const;
   bool isClangScope() const;
 
-  bool isCXX11Attribute() const { return SyntaxUsed == AS_CXX11 || IsAlignas; }
-
+  bool isAlignas() const { return IsAlignas; }
+  bool isCXX11Attribute() const { return SyntaxUsed == AS_CXX11; }
```

Comment:
This function was intended to tell you whether the attribute is a C++ attribute, and now it's changed to whether it's a C++ attribute spelled with `[[]]`. https://eel.is/c++draft/dcl.attr - note that `alignas` is an `attribute-specifier`. In C, `alignas` is *not* an attribute specifier (but it might become one someday) but is a kind of `declaration-specifier` instead.

I think it'd be better to keep this distinction, but I'll defer to @erichkeane if he has other opinions.

Reply:
I can see this both ways. It is typically used AFAIK as a 'is this a C++ spelling' despite not exactly meaning that.  On the other hand, as Aaron mentions, this is changing its meaning fairly significantly.  I'm still on the fence here.

Reply:
I realize there are backwards compatibility issues now (which I did not see before). Happy to defer to experts the decision.  I mostly went for a symptomatic treatment after I got failures since `kw_Alignas` was not added. Would it make sense to decompose `IsAlignAs` to `IsCXXAlignas || isCAlignAs` then? It'll bloat the code a bit but will allow retaining past behaviour by retaining the fine-grained information.

This is within an instance of an `AttributeCommonInfo` class and therefore an attribute, so I reasoned this is already an attribute - now these methods are intended to distinguish CXX or C is based on syntax.

Reply:
I think the trouble is that we've got code like this: https://github.com/llvm/llvm-project/blob/eb14f47bf1ccfda500ba3c3092d70e269f6f0b56/clang/lib/Parse/ParseDeclCXX.cpp#L3062 where we're using "isCXX11Attribute()" to mean "attribute-specifier" as a grammar term and not `[[]]` syntax specifically.

As best I can tell, I think we're fine for `isAlignas()` to report true for both `alignas` and `_Alignas` spellings; we can distinguish between the C and C++ spellings by looking at the language options in force. WDYT @erichkeane?

Reply:
Trying to guess based on surroundings, the `AttributeCommonInfo::is<Method>(...)` corresponds consistently to [`Syntax`](https://github.com/llvm/llvm-project/blob/4345baaa02c5e2d27a727e8b2cf22d95b2842cd9/clang/include/clang/Basic/AttributeCommonInfo.h#L26-L27) enum save `isCXX11Attribute`, and assuming consistency is preferred a method with the name `isCXX11Attribute` should only tell if it corresponds to `AS_CXX11`, which is `[[ ]]` when C++ is used (consistent with `is{C23,GNU,Declspec,Microsoft,...}Attribute()`).

This function body appears to imply what is defined as `attribute-specifier` ([C++](https://eel.is/c++draft/dcl.attr#nt:attribute-specifier)) prior to this change. So it may be worth renaming this and such intended usages to `isCXXAttributeSpecifier() = [[]] | alignas(..) ` which can be inferred from this instance to improve.  

In the [C23-draft](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3096.pdf) I picked off the Wikipedia page, `alignas` appears to be a _keyword_ part of an `alignment-specifier`, while `attribute-specifier` only includes `[[ ]]`. Common ancestor for `attribute-specifier`, `alignment-specifier` appears to be something called a `specifier-qualifier-list`. Based on the below details, I have a feeling it is okay to treat this as `isCAttributeSpecifier() == [[ ]] | alignas | _Alignas ` for practical purposes, moving the term `alignment-specifier` around. 

<details> <summary> C (grammar) </summary>

```
specifier-qualifier-list:
    type-specifier-qualifier attribute-specifier-sequence_opt
    type-specifier-qualifier specifier-qualifier-list
    
attribute-specifier-sequence:
    attribute-specifier-sequence_opt attribute-specifier

attribute-specifier:
    [[ attribute-list ]] 
    
type-specifier-qualifier:
    type-specifier
    type-qualifier
    alignment-specifier
    
alignment-specifier:
    alignas ( type-name )
    alignas ( constant-expression )
```
</details>